### PR TITLE
[rush] Remove chokidar dependency

### PR DESCRIPTION
**Pull request by dmichon-msft**
## Summary
The `chokidar` package provides a bunch of signal normalization that is redundant for the watch capabilities of rush, because the actual set of changed files is used purely as a boolean signal that is then handed off to the `PackageChangeAnalyzer` to determine what actually changed.

We already use `PackageChangeAnalyzer` (which is powered by `git diff` and is very, very fast) to determine the answer to "what changed" between incremental builds; as such, we prefer to also leverage it for the diffing in the watch case for consistent behavior.

This PR removes `chokidar` and replaces it with native `fs.watch` calls (carefully scoped to only watch directories), such that we can get a basic "hey, something (anything) changed, go poke git" notification as simply as possible.

## How it was tested
Local verification that a custom command using watch mode still behaves as expected.

### CODE REVIEW

File: common/config/rush/experiments.json [hunk 1]
```diff
@@ -18,7 +18,13 @@
...
    * Set this option to true to pass '--frozen-lockfile' instead.
    */
-  // "usePnpmFrozenLockfileForRushInstall": true,
+  "usePnpmFrozenLockfileForRushInstall": true,
+
+  /**
+   * By default, 'rush update' passes --no-prefer-frozen-lockfile to 'pnpm install'.
+   * Set this option to true to pass '--prefer-frozen-lockfile' instead to minimize shrinkwrap changes.
+   */
+  "usePnpmPreferFrozenLockfileForRushUpdate": true,
```
Comment (iclanton) [inline common/config/rush/experiments.json:27]: Unrelated?

File: apps/rush-lib/src/logic/ProjectWatcher.ts [hunk 2]
```diff
@@ -197,4 +271,29 @@ export class ProjectWatcher {
...
+    const rootSlashIndex: number = path.indexOf('/', projectRelativeFolder.length + 2);
+
+    if (rootSlashIndex < 0) {
+      yield path;
+      return;
+    }
+
+    yield path.slice(0, rootSlashIndex);
+
+    if (isRecursiveSupported) {
```
Comment (iclanton) [inline apps/rush-lib/src/logic/ProjectWatcher.ts:289]: This seems backwards.

File: apps/rush-lib/src/logic/ProjectWatcher.ts [hunk 3]
```diff
@@ -54,29 +54,38 @@ export class ProjectWatcher {
...
-      interval: 1000
-    });
-
-    // Only watch for changes in the requested project folders
+    const previousState: PackageChangeAnalyzer = initialChangeResult.state;
+    const repoRoot: string = Path.convertToSlashes(this._rushConfiguration.rushJsonFolder);
+
+    const pathsToWatch: Set<string> = new Set();
+
+    const isRecursiveSupported: boolean = os.platform() === 'win32';
```
Comment (iclanton) [inline apps/rush-lib/src/logic/ProjectWatcher.ts:70]: Oh does this mean "is recursive folder watch supported"" Maybe we can clear up the name.

File: common/config/rush/experiments.json (hunk 1 above)
Comment (dmichon-msft) [reply common/config/rush/experiments.json:27]: I turned this on because it was bothering me that rush didn't think anything changed when I dropped the dependency. That was probably actually a rush bug, though. I'll exclude this, it should be its own PR.

File: apps/rush-lib/package.json [hunk 4]
```diff
@@ -30,7 +30,6 @@
     "@rushstack/ts-command-line": "workspace:*",
```
Comment (octogonz) [inline apps/rush-lib/package.json:30]: > ## Summary
> 
> The `chokidar` package provides a bunch of signal normalization that is redundant for the watch capabilities of rush, because the actual set of changed files is used purely as a boolean signal that is then handed off to the `PackageChangeAnalyzer` to determine what actually changed.
> 
> This PR removes `chokidar` and replaces it with native `fs.watch` calls.

At HBO we originally used `fs.watch` for watching and found it to be extremely slow.  As in, flaming MacBooks twirling and flying off of people's desks like a Class C aerial spinner firework.  😋

[chokidar](https://www.npmjs.com/package/chokidar) implements some important optimizations and is the gold standard for file watching.  Are we sure we shouldn't be using it?

Comment (dmichon-msft) [reply apps/rush-lib/package.json:30]: The only thing that might matter to us in `chokidar` is using `fsevents` directly instead of `fs.watch` on MacOS. Every other "optimization" is completely counterproductive for our use case, since the file system watcher is *not* the authority on what files changed, which is normally what `chokidar` adds these various layers to do. Per official node documentation, NodeJS `fs.watch` uses FSEvents when watching directories, which are the only thing we are watching.

I'd be open to a performance comparison between this build and the one with chokidar on MacOS, but my expectation is that this version will actually be faster, since it is doing less work. Also gets rid of the annoying optional dependency error during installation on non-MacOS.

Comment (octogonz) [reply apps/rush-lib/package.json:30]: It would be smart to do some kind of measurements on Linux/Windows/Mac to double-check that `fs.watch` isn't regressing performance somehow.  Specifically with regards to [this optimization](https://www.npmjs.com/package/chokidar):

> On MacOS, chokidar by default uses a native extension exposing the Darwin FSEvents API. This provides very efficient recursive watching compared with implementations like kqueue available on most *nix platforms.

In the longer term we're anticipating a more scalable design where a centralized Rush process watches the entire monorepo folder tree, and then Jest/Webpack/etc get notified by Rush (rather than watching for file events themselves).  So although Rush's watching requirements are somewhat limited right now, it made me think that we may need more of Chokidar's advanced features later down the road.  

However, @dmichon-msft has pointed out that Rush's primary optimization is to use PackageChangeAnalyzer to leverage Git's change detection.  The `fs.watch` part only needs to provide rough granularity notifications which trigger the PackageChangeAnalyzer.  That sounds like a persuasive argument that we can safely eliminate Chokidar. 👍

**Comment by octogonz** [comment]
Keytar was eliminated in #2647, but installing Rush still prints some warnings:
```
C:\>npm install -g @microsoft/rush

npm WARN deprecated @opentelemetry/types@0.2.0: Package renamed to @opentelemetry/api, see https://github.com/open-telemetry/opentelemetry-js
npm WARN deprecated fsevents@2.1.3: "Please update to latest v2.3 or v2.2"

C:\Program Files\nodejs\rush -> C:\Program Files\nodejs\node_modules\@microsoft\rush\bin\rush
C:\Program Files\nodejs\rushx -> C:\Program Files\nodejs\node_modules\@microsoft\rush\bin\rushx

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.1.2 (node_modules\@microsoft\rush\node_modules\chokidar\node_modules\fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})

+ @microsoft/rush@5.46.0
added 10 packages from 3 contributors, removed 70 packages and updated 25 packages in 14.618s
```
Will this PR eliminate any of these warnings? (All of them?)

**Comment by dmichon-msft** [comment]
> Keytar was eliminated in #2647, but installing Rush still prints some warnings:
> 
> ```
> C:\>npm install -g @microsoft/rush
> 
> npm WARN deprecated @opentelemetry/types@0.2.0: Package renamed to @opentelemetry/api, see https://github.com/open-telemetry/opentelemetry-js
> npm WARN deprecated fsevents@2.1.3: "Please update to latest v2.3 or v2.2"
> 
> C:\Program Files\nodejs\rush -> C:\Program Files\nodejs\node_modules\@microsoft\rush\bin\rush
> C:\Program Files\nodejs\rushx -> C:\Program Files\nodejs\node_modules\@microsoft\rush\bin\rushx
> 
> npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.1.2 (node_modules\@microsoft\rush\node_modules\chokidar\node_modules\fsevents):
> npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
> 
> + @microsoft/rush@5.46.0
> added 10 packages from 3 contributors, removed 70 packages and updated 25 packages in 14.618s
> ```
> 
> Will this PR eliminate any of these warnings? (All of them?)

Should remove the `fsevents` warnings.

**Review by octogonz — APPROVED** [review]
